File: cloudflare/src/pipeline/dedupe.ts
import { NormalizedItem } from '../types';
import { findDuplicates } from '../db';
import { Env } from '../types';

export async function dedupeItems(items: NormalizedItem[], env: Env): Promise<NormalizedItem[]> {
  if (!items.length) return items;

  const urls = items.map((i) => i.url).filter(Boolean);
  const hashes = items.map((i) => i.content_hash).filter((h): h is string => !!h);

  const { byUrl, byHash } = await findDuplicates(env, urls, hashes);

  const urlToId = new Map<string, string>();
  for (const row of byUrl) {
    urlToId.set(row.url, row.id);
    if (row.canonical_url) urlToId.set(row.canonical_url, row.id);
  }
  const hashToId = new Map<string, string>();
  for (const row of byHash) hashToId.set(row.content_hash, row.id);

  // Within-batch dedup
  const seenUrls = new Map<string, number>();
  const seenHashes = new Map<string, number>();

  for (let i = 0; i < items.length; i++) {
    const item = items[i];
    const url = item.url;
    const canonical = item.canonical_url ?? url;
    const hash = item.content_hash;

    // Check DB
    const dupId = urlToId.get(url) ?? urlToId.get(canonical) ?? (hash ? hashToId.get(hash) : undefined);
    if (dupId) {
      item.is_duplicate = true;
      item.duplicate_of_id = dupId;
      continue;
    }

    // Check within batch
    if (url && seenUrls.has(url)) { item.is_duplicate = true; continue; }
    if (hash && seenHashes.has(hash)) { item.is_duplicate = true; continue; }

    if (url) seenUrls.set(url, i);
    if (canonical && canonical !== url) seenUrls.set(canonical, i);
    if (hash) seenHashes.set(hash, i);
  }

  return items;
}

File: src/pipeline/dedupe.py
from __future__ import annotations

from typing import Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.models.content import ContentItem


async def check_duplicates(
    items: list[dict[str, Any]],
    session: AsyncSession,
) -> list[dict[str, Any]]:
    """Check items for duplicates against existing content in DB.

    Marks duplicates by setting is_duplicate=True and duplicate_of_id.
    Uses both URL and content_hash for matching.
    """
    if not items:
        return items

    # Collect URLs and hashes for batch lookup
    urls = [i["url"] for i in items if i["url"]]
    hashes = [i["content_hash"] for i in items if i["content_hash"]]

    existing_by_url: dict[str, str] = {}
    existing_by_hash: dict[str, str] = {}

    if urls:
        result = await session.execute(
            select(ContentItem.id, ContentItem.url, ContentItem.canonical_url)
            .where(ContentItem.url.in_(urls) | ContentItem.canonical_url.in_(urls))
        )
        for row in result:
            existing_by_url[row.url] = row.id
            if row.canonical_url:
                existing_by_url[row.canonical_url] = row.id

    if hashes:
        result = await session.execute(
            select(ContentItem.id, ContentItem.content_hash)
            .where(ContentItem.content_hash.in_(hashes))
            .where(ContentItem.is_duplicate == False)  # noqa: E712
        )
        for row in result:
            if row.content_hash:
                existing_by_hash[row.content_hash] = row.id

    # Also dedupe within the current batch
    seen_urls: dict[str, int] = {}
    seen_hashes: dict[str, int] = {}

    for idx, item in enumerate(items):
        url = item["url"]
        canonical = item.get("canonical_url", url)
        h = item["content_hash"]

        # Check against DB
        dup_id = existing_by_url.get(url) or existing_by_url.get(canonical) or existing_by_hash.get(h)
        if dup_id:
            item["is_duplicate"] = True
            item["duplicate_of_id"] = dup_id
            continue

        # Check within batch
        if url and url in seen_urls:
            item["is_duplicate"] = True
            continue
        if h and h in seen_hashes:
            item["is_duplicate"] = True
            continue

        if url:
            seen_urls[url] = idx
        if h:
            seen_hashes[h] = idx

    return items
